core/bindings.rs

```diff
@@ -0,0 +1,88 @@
+use crate::isolate::Buf;
+use crate::libdeno::PinnedBuf;
+use futures::Future;
+use std::fmt;
+use std::io;
+
+pub type BindingResult<T> = std::result::Result<T, BindingError>;
+
+#[derive(Debug)]
+pub struct BindingError {
+  repr: Repr,
+}
+
+#[derive(Debug)]
+enum Repr {
+  Simple(String),
+  IoErr(io::Error),
+}
+
+pub fn new_binding_error(msg: String) -> BindingError {
+  BindingError {
+    repr: Repr::Simple(msg),
+  }
+}
+
+impl fmt::Display for BindingError {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    match self.repr {
+      Repr::Simple(ref err_str) => f.pad(err_str),
+      Repr::IoErr(ref err) => err.fmt(f),
+    }
+  }
+}
+
+impl std::error::Error for BindingError {
+  fn description(&self) -> &str {
+    match self.repr {
+      Repr::Simple(ref msg) => msg.as_str(),
+      Repr::IoErr(ref err) => err.description(),
+    }
+  }
+
+  fn cause(&self) -> Option<&dyn std::error::Error> {
+    match self.repr {
+      Repr::Simple(ref _msg) => None,
+      Repr::IoErr(ref err) => Some(err),
+    }
+  }
+}
+
+impl From<io::Error> for BindingError {
+  #[inline]
+  fn from(err: io::Error) -> Self {
+    Self {
+      repr: Repr::IoErr(err),
+    }
+  }
+}
+
+pub type BindingOpAsyncFuture =
+  Box<dyn Future<Item = Buf, Error = BindingError> + Send>;
+
+pub enum BindingOpSyncOrAsync {
```

Comment:
Let's make Op in isolate.rs be generic on the error and use it instead of this new enum?